ctdb/doc/ltdbtool.1
'\" t
.\"     Title: ltdbtool
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 05/04/2011
.\"    Manual:  
.\"    Source:  
.\"  Language: English
.\"
.TH "LTDBTOOL" "1" "05/04/2011" "" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ltdbtool \- handle ctdb\*(Aqs local tdb copies
.SH "SYNOPSIS"
.HP \w'\fBltdbtool\ [OPTIONS]\ COMMAND\fR\ 'u
\fBltdbtool [OPTIONS] COMMAND\fR
.SH "DESCRIPTION"
.PP
ltdbtool is a utility to cope with ctdb\*(Aqs local tdb copies (LTDBs) without connecting to a ctdb daemon\&.
.PP
It can be used to
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
dump the contents of a LTDB, optionally printing the ctdb record header information,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
convert between an LTDB and a non\-clustered tdb by adding or removing ctdb headers and
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
convert between 64 and 32 bit LTDBs where the ctdb record headers differ by 4 bytes of padding\&.
.RE
.sp
.SH "COMMANDS"
.PP
help
.RS 4
Print a help text\&.
.RE
.PP
dump <IDB>
.RS 4
Dump the contents of a LTDB file to standard output in a human\-readable format\&.
.RE
.PP
convert <IDB> <ODB>
.RS 4
Make a copy of a LTDB optionally adding or removing ctdb headers\&.
.RE
.SH "OPTIONS"
.PP
\-p
.RS 4
Dump with header information, similar to "ctdb catdb"\&.
.RE
.PP
\-s {0|32|64}
.RS 4
Specify how to determine the ctdb record header size for the input database:
.PP
0
.RS 4
no ctdb header
.RE
.PP
32
.RS 4
ctdb header size of a 32 bit system (20 bytes)
.RE
.PP
64
.RS 4
ctdb header size of a 64 bit system (24 bytes)
.RE
.sp
The default is 32 or 64 depending on the system architecture\&.
.RE
.PP
\-o {0|32|64}
.RS 4
Specify how to determine the ctdb record header size for the output database, see \-s
.RE
.PP
\-S <SIZE>
.RS 4
Explicitly specify the ctdb record header size of the input database in bytes\&.
.RE
.PP
\-O <SIZE>
.RS 4
Explicitly specify the ctdb record header size for the output database in bytes\&.
.RE
.PP
\-h
.RS 4
Print a help text\&.
.RE
.SH "EXAMPLES"
.PP
Print a local tdb in "tdbdump" style:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool dump idmap2\&.tdb\&.0
    
.fi
.if n \{\
.RE
.\}
.PP
Print a local tdb with header information similar to "ctdb catdb":
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool dump \-p idmap2\&.tdb\&.0
    
.fi
.if n \{\
.RE
.\}
.PP
Strip the ctdb headers from records:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool convert \-o0 idmap2\&.tdb\&.0 idmap\&.tdb
    
.fi
.if n \{\
.RE
.\}
.PP
Strip 64 bit ctdb headers from records, running on i386:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool convert \-s64 \-o0 idmap2\&.tdb\&.0 idmap\&.tdb
    
.fi
.if n \{\
.RE
.\}
.PP
Strip the ctdb headers from records by piping through tdbrestore:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool dump idmap2\&.tdb\&.0 | tdbrestore idmap\&.tdb
    
.fi
.if n \{\
.RE
.\}
.PP
Convert a local tdb from a 64 bit system for usage on a 32 bit system:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool convert \-s64 \-o32 idmap2\&.tdb\&.0 idmap2\&.tdb\&.1
    
.fi
.if n \{\
.RE
.\}
.PP
Add a default header:
.sp
.if n \{\
.RS 4
.\}
.nf
      ltdbtool convert \-s0 idmap\&.tdb idmap2\&.tdb\&.0
    
.fi
.if n \{\
.RE
.\}
.SH "SEE ALSO"
.PP
ctdbd(1), ctdb(1), tdbdump(1), tdbrestore(1),
\m[blue]\fB\%http://ctdb.samba.org/\fR\m[]
.SH "COPYRIGHT/LICENSE"
.sp
.if n \{\
.RS 4
.\}
.nf
Copyright (C) Gregor Beck 2011
Copyright (C) Michael Adam 2011

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or (at
your option) any later version\&.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&.  See the GNU
General Public License for more details\&.

You should have received a copy of the GNU General Public License
along with this program; if not, see http://www\&.gnu\&.org/licenses/\&.
.fi
.if n \{\
.RE
.\}


NEWS
Release Notes - Heimdal - Version Heimdal 1.3

 New features

 - Partital support for MIT kadmind rpc protocol in kadmind
 - Better support for finding keytab entries when using SPN aliases in the KDC
 - Support BER in ASN.1 library (needed for CMS)
 - Support decryption in Keychain private keys
 - Support for new sqlite based credential cache
 - Try both to KDC referals the the common DNS reverse lookup in GSS-API
 - Fix the KCM not not leak resources on failure
 - Add IPv6 support to iprop
 - Support localization of error strings in
   kinit/klist/kdestroy and Kerberos library
 - Remove Kerberos 4 support in application (still in KDC)
 - Deprecate DES
 - Support i18n password in windows domains (using UTF-8)
 - More complete API emulation of OpenSSL in hcrypto
 - Support for ECDSA and ECDH when linking with OpenSSL

 API changes

 - Support for settin friendly name on credential caches
 - Move to using doxygen to generate documentation.
 - Sprinkling __attribute__((depricated)) for old function to be removed
 - Support to export LAST-REQUST information in AS-REQ
 - Support for client deferrals in in AS-REQ
 - Add seek support for krb5_storage.
 - Support for split AS-REQ, first step for IA-KERB
 - Fix many memory leaks and bugs
 - Improved regression test
 - Support krb5_cccol
 - Switch to krb5_set_error_message
 - Support krb5_crypto_*_iov	
 - Switch to use EVP for most function
 - Use SOCK_CLOEXEC and O_CLOEXEC (close on exec)
 - Add support for GSS_C_DELEG_POLICY_FLAG
 - Add krb5_cc_[gs]et_config to store data in the credential caches
 - PTY testing application

Bugfixes
 - Make building on AIX6 possible.
 - Bugfixes in LDAP KDC code to make it more stable
 - Make ipropd-slave reconnect when master down gown


Release Notes - Heimdal - Version Heimdal 1.2.1

* Bug

  [HEIMDAL-147] - Heimdal 1.2 not compiling on Solaris
  [HEIMDAL-151] - Make canned tests work again after cert expired
  [HEIMDAL-152] - iprop test: use full hostname to avoid realm
                  resolving errors
  [HEIMDAL-153] - ftp: Use the correct length for unmap, msync

Release Notes - Heimdal - Version Heimdal 1.2

* Bug

  [HEIMDAL-10] - Follow-up on bug report for SEGFAULT in
  		 gss_display_name/gss_export_name when using SPNEGO
  [HEIMDAL-15] - Re: [Heimdal-bugs] potential bug in Heimdal 1.1
  [HEIMDAL-17] - Remove support for depricated [libdefaults]capath
  [HEIMDAL-52] - hdb overwrite aliases for db databases
  [HEIMDAL-54] - Two issues which affect credentials delegation
  [HEIMDAL-58] - sockbuf.c calls setsockopt with bad args
  [HEIMDAL-62] - Fix printing of sig_atomic_t
  [HEIMDAL-87] - heimdal 1.1 not building under cygwin in hcrypto
  [HEIMDAL-105] - rcp: sync rcp with upstream bsd rcp codebase
  [HEIMDAL-117] - Use libtool to detect symbol versioning (Debian Bug#453241)

* Improvement
  [HEIMDAL-67] - Fix locking and store credential in atomic writes
                 in the FILE credential cache
  [HEIMDAL-106] - make compile on cygwin again
  [HEIMDAL-107] - Replace old random key generation in des module
                  and use it with RAND_ function instead
  [HEIMDAL-115] - Better documentation and compatibility in hcrypto
                  in regards to OpenSSL

* New Feature
  [HEIMDAL-3] - pkinit alg agility PRF test vectors
  [HEIMDAL-14] - Add libwind to Heimdal
  [HEIMDAL-16] - Use libwind in hx509
  [HEIMDAL-55] - Add flag to krb5 to not add GSS-API INT|CONF to
                 the negotiation
  [HEIMDAL-74] - Add support to report extended error message back
                 in AS-REQ to support windows clients
  [HEIMDAL-116] - test pty based application (using rkpty)
  [HEIMDAL-120] - Use new OpenLDAP API (older deprecated)

* Task
  [HEIMDAL-63] - Dont try key usage KRB5_KU_AP_REQ_AUTH for TGS-REQ.
                 This drop compatibility with pre 0.3d KDCs.
  [HEIMDAL-64] - kcm: first implementation of kcm-move-cache
  [HEIMDAL-65] - Failed to compile with --disable-pk-init
  [HEIMDAL-80] - verify that [VU#162289]: gcc silently discards some
                 wraparound checks doesn't apply to Heimdal

Changes in release 1.1

 * Read-only PKCS11 provider built-in to hx509.

 * Documentation for hx509, hcrypto and ntlm libraries improved.

 * Better compatibilty with Windows 2008 Server pre-releases and Vista.

 * Mac OS X 10.5 support for native credential cache.

 * Provide pkg-config file for Heimdal (heimdal-gssapi.pc).

 * Bug fixes.

Changes in release 1.0.2

* Ubuntu packages.

* Bug fixes.

Changes in release 1.0.1

 * Serveral bug fixes to iprop.

 * Make work on platforms without dlopen.

 * Add RFC3526 modp group14 as default.

 * Handle [kdc] database = { } entries without realm = stanzas.

 * Make krb5_get_renewed_creds work.

 * Make kaserver preauth work again.

 * Bug fixes.

Changes in release 1.0

 * Add gss_pseudo_random() for mechglue and krb5.

 * Make session key for the krbtgt be selected by the best encryption
   type of the client.

 * Better interoperability with other PK-INIT implementations.

 * Inital support for Mac OS X Keychain for hx509.

 * Alias support for inital ticket requests.

 * Add symbol versioning to selected libraries on platforms that uses
   GNU link editor: gssapi, hcrypto, heimntlm, hx509, krb5, and libkdc.

 * New version of imath included in hcrypto.

 * Fix memory leaks.

 * Bugs fixes.

Changes in release 0.8.1

 * Make ASN.1 library less paranoid to with regard to NUL in string to
   make it inter-operate with MIT Kerberos again.

 * Make GSS-API library work again when using gss_acquire_cred

 * Add symbol versioning to libgssapi when using GNU ld.

 * Fix memory leaks 

 * Bugs fixes

Changes in release 0.8

 * PK-INIT support.

 * HDB extensions support, used by PK-INIT.

 * New ASN.1 compiler.

 * GSS-API mechglue from FreeBSD.

 * Updated SPNEGO to support RFC4178.

 * Support for Cryptosystem Negotiation Extension (RFC 4537).

 * A new X.509 library (hx509) and related crypto functions.


ctdb/tools/ltdbtool.c
/*
 * ctdb local tdb tool
 *
 * Copyright (C) Gregor Beck 2011
 * Copyright (C) Michael Adam 2011
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <ctype.h> /* isprint */
#include <string.h> /* strstr */
#include <fcntl.h> /* mode_t */
#include <sys/stat.h> /* S_IRUSR */
#include <stdint.h> /* uint32_t */
#include <netinet/in.h> /* struct sockaddr_in */
#include <sys/socket.h> /* struct sockaddr */
#include <sys/param.h>  /* MIN */
#include <tdb.h>
#include <unistd.h> /* getopt */
#include <errno.h>

#include "ctdb_protocol.h"

enum {
	MAX_HEADER_SIZE=24,
	OUT_MODE = S_IRUSR | S_IWUSR,
	OUT_FLAGS = O_EXCL|O_CREAT|O_RDWR,
};

union  ltdb_header {
	struct ctdb_ltdb_header hdr;
	uint32_t uints[MAX_HEADER_SIZE/4];
};

static const union ltdb_header DEFAULT_HDR = {
	.hdr.dmaster = -1,
};

static int help(const char* cmd)
{
	fprintf(stdout, ""
"Usage: %s [options] <command>\n"
"\n"
"Options:\n"
"   -s {0|32|64}    specify how to determine the ctdb record header size\n"
"                   for the input database:\n"
"                   0: no ctdb header\n"
"                   32: ctdb header size of a 32 bit system (20 bytes)\n"
"                   64: ctdb header size of a 64 bit system (24 bytes)\n"
"                   default: 32 or 64 depending on the system architecture\n"
"\n"
"   -S <num>        the number of bytes to interpret as ctdb record header\n"
"                   for the input database (beware!)\n"
"\n"
"   -o {0|32|64}    specify how to determine the ctdb record header size\n"
"                   for the output database\n"
"                   0: no ctdb header\n"
"                   32: ctdb header size of a 32 bit system (20 bytes)\n"
"                   64: ctdb header size of a 64 bit system (24 bytes)\n"
"                   default: 32 or 64 depending on the system architecture\n"
"\n"
"   -O <num>        the number of bytes to interpret as ctdb record header\n"
"                   for the output database (beware!)\n"
"\n"
"   -p              print header (for the dump command), defaults ot off\n"
"\n"
"   -h              print this help\n"
"\n"
"Commands:\n"
"  help                         print this help\n"
"  dump <db>                    dump the db to stdout\n"
"  convert <in_db> <out_db>     convert the db\n\n", cmd);
	return 0;
}

static int usage(const char* cmd)
{
	fprintf(stderr,
		"Usage: %s dump [-p] [-s{0|32|64}] <idb>\n"
		"       %s convert [-s{0|32|64}] [-o{0|32|64}] <idb> <odb>\n"
		"       %s {help|-h}\n"
		, cmd, cmd, cmd);
	return -1;
}

static int
ltdb_traverse(TDB_CONTEXT *tdb, int (*fn)(TDB_CONTEXT*, TDB_DATA, TDB_DATA,
					  struct ctdb_ltdb_header*, void *),
	      void *state, int hsize);

struct write_record_ctx {
	TDB_CONTEXT* tdb;
	size_t hsize;
	int tdb_store_flags;
};

static int
write_record(TDB_CONTEXT* tdb, TDB_DATA key, TDB_DATA val,
	     struct ctdb_ltdb_header* hdr,
	     void* write_record_ctx);


struct dump_record_ctx {
	FILE* file;
	void (*print_data)(FILE*, TDB_DATA);
	void (*dump_header)(struct dump_record_ctx*, struct ctdb_ltdb_header*);
};

static int dump_record(TDB_CONTEXT* tdb, TDB_DATA key, TDB_DATA val,
		       struct ctdb_ltdb_header* hdr,
		       void* dump_record_ctx);
static void print_data_tdbdump(FILE* file, TDB_DATA data);
static void dump_header_full(struct dump_record_ctx*, struct ctdb_ltdb_header*);
static void dump_header_nop(struct dump_record_ctx* c,
			    struct ctdb_ltdb_header* h)
{}

static int dump_db(const char* iname, FILE* ofile, int hsize, bool dump_header)
{
	int ret = -1;
	TDB_CONTEXT* idb = tdb_open(iname, 0, TDB_DEFAULT, O_RDONLY, 0);
	if (!idb) {
		perror("tdbopen in");
	} else {
		struct dump_record_ctx dump_ctx = {
			.file = ofile,
			.print_data =  &print_data_tdbdump,
			.dump_header = dump_header ? &dump_header_full
			                           : &dump_header_nop,
		};
		ret = ltdb_traverse(idb, &dump_record, &dump_ctx, hsize);
		tdb_close(idb);
	}
	return ret;
}

static int conv_db(const char* iname, const char* oname, size_t isize, size_t osize)
{
	int ret = -1;
	TDB_CONTEXT* idb = tdb_open(iname, 0, TDB_DEFAULT, O_RDONLY, 0);
	if (!idb) {
		perror("tdbopen in");
	} else {
		TDB_CONTEXT* odb = tdb_open(oname, 0, TDB_DEFAULT, OUT_FLAGS, OUT_MODE);
		if (!odb) {
			perror("tdbopen out");
		} else {
			struct write_record_ctx ctx = {
				.tdb = odb,
				.hsize = osize,
				.tdb_store_flags = TDB_REPLACE,
			};
			ret = ltdb_traverse(idb, &write_record, &ctx, isize);
			tdb_close(odb);
		}
		tdb_close(idb);
	}
	return ret;
}

static bool parse_size(size_t* size, const char* arg, bool raw) {
	long val;
	errno = 0;
	val = strtol(arg, (char **) NULL, 10);
	if (errno != 0) {
		return false;
	}
	if (!raw) {
		switch(val) {
		case 0:
			break;
		case 32:
			val = 20;
			break;
		case 64:
			val = 24;
			break;
		default:
			return false;
		}
	}
	*size = MIN(val, MAX_HEADER_SIZE);
	return true;
}


int main(int argc, char* argv[])
{
	size_t isize = sizeof(struct ctdb_ltdb_header);
	size_t osize = sizeof(struct ctdb_ltdb_header);
	bool print_header = false;
	int opt;
	const char *cmd, *idb, *odb;

	while ((opt = getopt(argc, argv, "s:o:S:O:ph")) != -1) {
		switch (opt) {
		case 's':
		case 'S':
			if (!parse_size(&isize, optarg, isupper(opt))) {
				return usage(argv[0]);
			}
			break;
		case 'o':
		case 'O':
			if (!parse_size(&osize, optarg, isupper(opt))) {
				return usage(argv[0]);
			}
			break;
		case 'p':
			print_header = true;
			break;
		case 'h':
			return help(argv[0]);
		default:
			return usage(argv[0]);
		}
	}

	if (argc - optind < 1) {
		return usage(argv[0]);
	}

	cmd = argv[optind];

	if (strcmp(cmd, "help") == 0) {
		return help(argv[0]);
	}
	else if (strcmp(cmd, "dump") == 0) {
		int ret;
		if (argc - optind != 2) {
			return usage(argv[0]);
		}
		idb = argv[optind+1];
		ret = dump_db(idb, stdout, isize, print_header);
		return (ret >= 0) ? 0 : ret;
	}
	else if (strcmp(cmd, "convert") == 0) {
		int ret;
		if (argc - optind != 3) {
			return usage(argv[0]);
		}
		idb = argv[optind+1];
		odb = argv[optind+2];
		ret = conv_db(idb, odb, isize, osize);
		return (ret >= 0) ? 0 : ret;
	}

	return usage(argv[0]);
}

struct ltdb_traverse_ctx {
	int (*fn)(TDB_CONTEXT*,TDB_DATA,TDB_DATA,struct ctdb_ltdb_header*,void *);
	void* state;
	size_t hsize;
};

static int
ltdb_traverse_fn(TDB_CONTEXT* tdb, TDB_DATA key, TDB_DATA val,
		 void* ltdb_traverse_ctx)
{
	struct ltdb_traverse_ctx* ctx =
		(struct ltdb_traverse_ctx*)ltdb_traverse_ctx;
	union ltdb_header hdr = DEFAULT_HDR;

	const size_t hsize = MIN(sizeof(hdr), ctx->hsize);
	if (val.dsize < hsize) {
		fprintf(stderr, "Value too short to contain a ctdb header: ");
		print_data_tdbdump(stderr, key);
		fprintf(stderr, " = ");
		print_data_tdbdump(stderr, val);
		fputc('\n', stderr);
		return -1;
	}

	memcpy(&hdr, val.dptr, hsize);

	if (hdr.uints[5] != 0) {
		fprintf(stderr, "Warning: header padding isn't zero! Wrong header size?\n");
	}
	val.dptr += ctx->hsize;
	val.dsize -= ctx->hsize;
	return ctx->fn(tdb, key, val, &hdr.hdr, ctx->state);
}

int ltdb_traverse(TDB_CONTEXT *tdb,
		  int (*fn)(TDB_CONTEXT *,TDB_DATA,TDB_DATA,struct ctdb_ltdb_header*,void *),
		  void *state, int hsize)
{
	struct ltdb_traverse_ctx ctx = {
		.fn = fn,
		.state = state,
		.hsize = hsize < 0 ? sizeof(struct ctdb_ltdb_header) : hsize,
	};
	return tdb_traverse(tdb, &ltdb_traverse_fn, &ctx);
}

int write_record(TDB_CONTEXT* tdb, TDB_DATA key, TDB_DATA val,
		 struct ctdb_ltdb_header* hdr,
		 void* write_record_ctx)
{
	struct write_record_ctx* ctx
		= (struct write_record_ctx*)write_record_ctx;

	if (ctx->hsize == 0) {
		if (tdb_store(ctx->tdb, key, val, ctx->tdb_store_flags) == -1) {
			fprintf(stderr, "tdb_store: %s\n", tdb_errorstr(ctx->tdb));
			return -1;
		}
	} else {
		TDB_DATA h = {
			.dptr = (void*)hdr,
			.dsize = ctx->hsize,
		};
		if(tdb_store(ctx->tdb, key, h, ctx->tdb_store_flags) == -1) {
			fprintf(stderr, "tdb_store: %s\n", tdb_errorstr(ctx->tdb));
			return -1;
		}
		if(tdb_append(ctx->tdb, key, val) == -1) {
			fprintf(stderr, "tdb_append: %s\n", tdb_errorstr(ctx->tdb));
			return -1;
		}
	}
	return 0;
}

int dump_record(TDB_CONTEXT* tdb, TDB_DATA key, TDB_DATA val,
		struct ctdb_ltdb_header* hdr,
		void* dump_record_ctx)
{
	struct dump_record_ctx* ctx = (struct dump_record_ctx*)dump_record_ctx;

	fprintf(ctx->file, "{\nkey(%d) = ", (int)key.dsize);
	ctx->print_data(ctx->file, key);
	fputc('\n', ctx->file);
	ctx->dump_header(ctx, hdr);
	fprintf(ctx->file, "data(%d) = ", (int)val.dsize);
	ctx->print_data(ctx->file, val);
	fprintf(ctx->file, "\n}\n");
	return 0;
}

void dump_header_full(struct dump_record_ctx* c, struct ctdb_ltdb_header* h)
{
	fprintf(c->file, "dmaster: %d\nrsn: %llu\nflags: 0x%X\n",
		(int)h->dmaster,
		(unsigned long long)h->rsn, h->flags);
}

void print_data_tdbdump(FILE* file, TDB_DATA data) {
	unsigned char *ptr = data.dptr;
	fputc('"', file);
	while (data.dsize--) {
		if (isprint(*ptr) && !strchr("\"\\", *ptr)) {
			fputc(*ptr, file);
		} else {
			fprintf(file, "\\%02X", *ptr);
		}
		ptr++;
	}
	fputc('"',file);
}

